ngraph_bridge/ngraph_data_cache.h

```diff
@@ -0,0 +1,139 @@
+// Executor Cache.h
+
+/*******************************************************************************
+ * Copyright 2019 Intel Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/
+
+#ifndef NGRAPH_DATA_CACHE_H_
+#define NGRAPH_DATA_CACHE_H_
+#pragma once
+
+#include <mutex>
+#include <ostream>
+#include <vector>
+#include "absl/synchronization/mutex.h"
+
+#include "tensorflow/core/framework/tensor_shape.h"
+#include "tensorflow/core/graph/graph.h"
+
+#include "logging/ngraph_log.h"
+#include "ngraph/ngraph.hpp"
+
+#include "ngraph_bridge/ngraph_freshness_tracker.h"
+#include "ngraph_bridge/ngraph_pipelined_tensors.h"
+
+namespace tensorflow {
+
+namespace ngraph_bridge {
+
+// Forward declaration for friend class
+namespace testing {
+class NGraphDataCacheTest_SameKeyMultiThread_Test;
+class NGraphDataCacheTest_DiffKeyMultiThread_Test;
+}
+
+template <typename T1, typename T2>
+class NgraphDataCache {
+ public:
+  NgraphDataCache(int depth);
+  ~NgraphDataCache();
+
+  // This method performs lookup in the cache for requested key, if not found
+  // it will create item, put it in the cache and returns item and stattus
+  std::pair<Status, T2> LookUpOrCreate(
+      T1 key, std::function<std::pair<Status, T2>()> create_item,
+      std::function<void(T2)> callback_destroy_item);
+  std::pair<Status, T2> LookUpOrCreate(
+      T1 key, std::function<std::pair<Status, T2>()> create_item);
+
```

Comment:
Need to provide the following functions:
1. An iterator so that the caller can browse the cache.
2. Add a `RemoveItem(key)` method that removes an item from the cache 
2.1. A variation that doesn't have any callback
2.2. A variation that takes a destroy callback 
3. Add a `RemoveAll(destroy_callback)`

These are necessary to enable deterministic resource release (e.g., backend tensors/executables can be predictably removed to reclaim the resources).